src/operator/activation-inl.h

```diff
@@ -82,7 +88,27 @@ class ActivationOp : public Operator {
     Tensor<xpu, 2> m_out_grad = out_grad[activation::kOut].FlatTo2D<xpu, real_t>(s);
     Tensor<xpu, 2> m_out_data = out_data[activation::kOut].FlatTo2D<xpu, real_t>(s);
     Tensor<xpu, 2> m_in_grad = in_grad[activation::kData].FlatTo2D<xpu, real_t>(s);
-    Assign(m_in_grad, req[activation::kData], F<BackwardOp>(m_out_data) * m_out_grad);
+    switch (param_.act_type) {
+    case activation::kReLU:
+      Assign(m_in_grad, req[activation::kData], F<mshadow_op::relu_grad>(m_out_data) * m_out_grad);
+      break;
+    case activation::kSigmoid:
+      Assign(m_in_grad, req[activation::kData],
+             F<mshadow_op::sigmoid_grad>(m_out_data) * m_out_grad);
+      break;
+    case activation::kTanh:
+      Assign(m_in_grad, req[activation::kData],
+             F<mshadow_op::tanh_grad>(m_out_data) * m_out_grad);
+      break;
+    case activation::kSoftReLU:
+      Assign(m_in_grad, req[activation::kData],
+             F<mshadow_op::softrelu_grad>(m_out_data) * m_out_grad);
+      break;
+    case activation::kSoftSign:
```

Comment:
Maybe we can reduce the code duplication, by just special casing `SoftSign`? Also by using `<type_traits>` and `is_same` we maybe don't need to introduce `param_`.